test/behavior/ptrcast.zig

```diff
@@ -91,14 +89,16 @@ test "reinterpret bytes of an extern struct into another" {
 
 fn testReinterpretExternStructAsExternStruct() !void {
     const S1 = extern struct {
-        a: u8,
-        b: u16,
+        a: u32,
+        b: u32,
         c: u8,
     };
-    comptime var bytes align(2) = S1{ .a = 0, .b = 0, .c = 5 };
+    comptime var bytes: S1 = .{ .a = 0, .b = 0, .c = 5 };
 
     const S2 = extern struct {
-        a: u32 align(2),
+        a0: u32,
+        a1: u16,
+        a2: u16,
```

Comment:
I don't see anything wrong with the previous version of this test case- it looks like valid Zig code to me. So this alteration of the behavior test is a regression. If you want to add additional coverage with this new case, that's fine. But if it's valid Zig code then it has to lower to valid C code in the C backend.

Same deal with the other altered behavior tests here.

Reply:
The challenge here is that the CBE can't support some Zig features, specifically under-aligned fields.

Constructing the field pointer in the other test is also UB according to ubsan, since the parent pointer does not fit in the bounds of the pointee object. We can potentially work around that by using `offsetof` and pointer arithmetic.

Should we skip these tests for the CBE indefinitely, and add others that are within the range of defined behavior for C?

Reply:
> The challenge here is that the CBE can't support some Zig features, specifically under-aligned fields.

That is incorrect, because C is turing complete. If we must lower some types as byte arrays then so be it.

There are also non-standard compiler extensions that we can take advantage of such as the [aligned attribute](https://gcc.gnu.org/onlinedocs/gcc-11.2.0/gcc/Common-Type-Attributes.html#Common-Type-Attributes).

> Constructing the field pointer in the other test is also UB according to ubsan, since the parent struct does not fit in the bounds of the pointee object.

The correct size of `S` is 6 bytes, the pointer cast is valid zig, and the C backend's job is to lower it, without invoking UB at the C language layer. I don't think a workaround with offsetof and pointer arithmetic is the way to go, instead, the C backend should lower the `S` type to something that is only 6 bytes.

> Should we skip these tests for the CBE indefinitely, and add others that are within the range of defined behavior for C?

No. We will bend C our will, and make them all pass.

Reply:
I see - A complete feature set is the hard requirement, and using canonical C features is simply best-effort :+1:

Is there any clear line about which compilers we are targeting and when we can use non-standard extensions?

FWIW, the field pointer problem shows up in a test like this:
```zig
const bytes align(4) = [_]u8{1, 2, 3, 4};
const S = struct {
    a: u16,
    b: u16,
    c: u32,
};

const x = @ptrCast(*const S, &bytes);
var p = &x.b; // triggers UB in the C output
_ = p;
```

I will add this test to the PR, so that we don't forget to support this in the CBE

Reply:
> Is there any clear line about which compilers we are targeting and when we can use non-standard extensions?

The goal is to target C99, and any non-standard extensions are fair game as long as we also support the extensions not being available (via preprocessor checks). You can see some of this in zig.h.